route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,238 @@ namespace route_following_plugin
+        auto nearest_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc_, 10);   //Return 10 nearest lanelets
+        if(nearest_lanelets.size() == 0)
         {
             ROS_WARN_STREAM("Cannot find any lanelet in map!");
-            return true;
+            return maneuvers; 
         }
-        
-        auto shortest_path = wm_->getRoute()->shortestPath();
 
-        lanelet::ConstLanelet current_lanelet;
-        int last_lanelet_index = -1;
-        for (auto llt : current_lanelets)
+        //Check if current position is on shortest path, if not on shortest path return empty maneuver  
+        int current_lanelet_index = -1;
+        for (auto llt : nearest_lanelets)
         {
-            if (boost::geometry::within(current_loc, llt.second.polygon2d()))
+            if (boost::geometry::within(current_loc_, llt.second.polygon2d()))
             {
-                int potential_index = findLaneletIndexFromPath(llt.second.id(), shortest_path);
+                int potential_index = findLaneletIndexFromPath(llt.second.id(), route_shortest_path);
                 if (potential_index != -1)
                 {
-                    last_lanelet_index = potential_index;
-                    current_lanelet = shortest_path[last_lanelet_index];
+                    current_lanelet_index = potential_index;
                     break;
                 }
             }
         }
-        if(last_lanelet_index == -1)
+        if(current_lanelet_index == -1)
         {
-            ROS_ERROR_STREAM("Current position is not on the shortest path! Returning an empty maneuver");
-            return true;
+            ROS_WARN_STREAM("Current position is not on the shortest path! Returning an empty maneuver");
         }
-        double current_progress = wm_->routeTrackPos(current_loc).downtrack;
-        double speed_progress = current_speed_;
-        ros::Time time_progress = ros::Time::now();
-        double target_speed=findSpeedLimit(current_lanelet);   //get Speed Limit
 
-        double total_maneuver_length = current_progress + mvr_duration_ * target_speed;
-        double route_length=  wm_->getRouteEndTrackPos().downtrack; 
-        total_maneuver_length = std::min(total_maneuver_length, route_length);
+        double route_length = wm_->getRouteEndTrackPos().downtrack;
+        double prev_mvr_end_speed = 0.0;
+        //Find lane changes in path - up to the second to last lanelet in path (till lane change is possible)
+        for(int i = 0; i < route_shortest_path.size() - 1; ++i){
+            auto following_lanelets = wm_->getRoute()->followingRelations(route_shortest_path[i]);
+            double target_speed_in_lanelet = findSpeedLimit(route_shortest_path[i]);
+            double start_dist = wm_->routeTrackPos(route_shortest_path[i].centerline2d().front()).downtrack;
+            double end_dist = wm_->routeTrackPos(route_shortest_path[i].centerline2d().back()).downtrack; 
```

Comment:
The start distance and start time should be assigned from the previous maneuver so that we avoid possibly small floating point deltas on lanelet downtrack calculations. There is actually no need to track these values as unique variables like you did with prev_maver_end_speed you can just extract them from maneuvers.back() on each loop.

Reply:
I think the fix still has a problem in it.

Reply:
@TonysCousin what are you referring too here?